akka-projection-core/src/main/scala/akka/projection/internal/InternalProjectionState.scala

```diff
@@ -118,25 +129,28 @@ private[akka] abstract class InternalProjectionState[Offset, Envelope](
         case single: SingleHandlerStrategy[Envelope] =>
           val handler = single.handler()
           val handlerRecovery =
-            HandlerRecoveryImpl[Offset, Envelope](projectionId, recoveryStrategy, logger, statusObserver)
+            HandlerRecoveryImpl[Offset, Envelope](projectionId, recoveryStrategy, logger, statusObserver, telemetry)
 
           Flow[ProjectionContextImpl[Offset, Envelope]].mapAsync(parallelism = 1) { context =>
+            val measured: () => Future[Done] = { () =>
+              handler.process(context.envelope).map { done =>
+                telemetry.afterProcess(context.nanosSinceReady())
+                done
+              }
+            }
             handlerRecovery
-              .applyRecovery(
-                context.envelope,
-                context.offset,
-                context.offset,
-                abort.future,
-                () => handler.process(context.envelope))
-              .map(_ => context)
+              .applyRecovery(context.envelope, context.offset, context.offset, abort.future, measured)
+              .map { _ =>
+                context
+              }
+
           }
 
         case grouped: GroupedHandlerStrategy[Envelope] =>
           val groupAfterEnvelopes = grouped.afterEnvelopes.getOrElse(settings.groupAfterEnvelopes)
           val groupAfterDuration = grouped.orAfterDuration.getOrElse(settings.groupAfterDuration)
-          val handler = grouped.handler()
```

Comment:
was there a reason for changing this? it was intended that the `handler()` should only be called once instead of for each envelope because it has now some lazy init overhead

Reply:
That's a mistake. While rebasing my branch a couple of these changed slipped and I thought I had reviewed them all. Good catch.